content/chapters/compute/lecture/slides.md

```diff
@@ -0,0 +1,1164 @@
+
+        /* Scheduler information */
+        struct sched_entity             se;
+        const struct sched_class        *sched_class;
+
+        /* The VAS: memory mappings */
+        struct mm_struct                *mm;
+
+        int                             exit_state;
+        int                             exit_code;
+
+        pid_t                           pid;
+
+        struct task_struct __rcu        *parent;
+
+        /* Child processes */
+        struct list_head                children;
+
+        /* Open file information */
+        struct files_struct             *files;
+};
+```
+
+---
+
+### `fork`
+
+* Spawns child process
+* Child process starts execution right after the call to `fork`
+* Returns twice
+    * `0` in child process
+    * Child's PID in parent process
+[Quiz](./quiz/process-creation.md)
+* Initially child and parent share the same PAS (Physical Address Space)
+* Child's writable pages are marked **copy-on-write**
+
+----
+
+### Copy-on-write
+
+When `fork` returns, the child inherits the parent's page table.
+
+![Copy-on-write after fork()](./media/copy-on-write-initial.svg)
+
+----
+
+### Copy-on-write
+
+* Writable pages are **copied and modified** when written: `demo/copy-on-write/copy_on_write_overhead.c`.
+* [Quiz](./quiz/sections-always-shared.md)
+
+![Copy-on-write after Write](./media/copy-on-write-final.svg)
+
+----
+
+### Copy-on-write Overheard
+
+* The overhead from writing data comes mostly from minor page faults
+```
+student@os:~/.../compute/lecture/demo/copy-on-write$ ./copy_on_write_overhead 
+ * Child prcess started
+ -- Press ENTER to continue ...
+Time for reading 131072 pages: 30 ms
+ * Child process read pages
+ -- Press ENTER to continue ...
+Time for writing to 131072 pages: 222 ms
+[...]
+```
+
+* The child process performs exactly 131072 **minor** page faults when writing data
+```
+student@os:~$ cat /proc/$(pidof copy_on_write_overhead)/stat | cut -d ' ' -f 10  # before writing
+22
+
+student@os:~$ cat /proc/$(pidof copy_on_write_overhead)/stat | cut -d ' ' -f 10  # after writing
+131094
+```
+
+---
+
+## 2. Threads - Backstory
+
+* You have an application that applies filters to a set of images
+* Processes need to share all images
+    * Difficult because the images (like all data) are marked copy-on-write
+* **Solution:** use threads
+    * No copy-on-write: threads share the heap and data sections
+
+----
+
+### 2. Threads - Definitions
+
+* Each process has at least one thread
+* Lightweight processes (LWP)
+* Scheduling unit
+
+---
+
+### Thread - Resources
+
+![Thread Resources](media/process-vs-thread.png)
+
+----
+
+### App, Process, Thread - The Bigger Picture
+
+* [Quiz](./quiz/threads-shared-data.md)
+
+![Application Architecture](media/app-process-thread.svg)
+
+----
+
+### `fork` and Threads
+
+* `fork` also creates a new thread...
+* ... but part of a different process
+
+![fork - new Thread](./media/fork-thread.svg)
+
+---
+
+### Process vs Thread
+
```

Comment:
I think most of the implementation effort is somewhat common. If you only had processes and not separate threads (i.e. a process only contains a single thread), you would run into the same issues. The problem actually comes from parallelism itself rather than threads or processes per se.